javascript/ql/src/Declarations/UnreachableMethodOverloads.ql

```diff
@@ -0,0 +1,131 @@
+     */
+    string getName() {
+        result = name
+    }
+}
+
+/**
+ * Holds if the two call call-signatures could be overloads of each other and have the same parameter types.
+ */
+predicate matchingCallSignature(MethodCallSig method, MethodCallSig other) {
+  method.getName() = other.getName() and 
+  
+  method.getNumOptionalParameter() = other.getNumOptionalParameter() and
+  method.getNumParameter() = other.getNumParameter() and
+  method.getNumRequiredParameter() = other.getNumRequiredParameter() and
+  // purposely not looking at number of type arguments. 
+  
+  method.getKind() = other.getKind() and 
+  
+  
+  forall(int i | i in [0 .. -1 + method.getNumParameter()] | 
+    method.getParameter(i) = other.getParameter(i) // This is sometimes imprecise, so it is still a good idea to compare type annotations.
+  ) and 
+  
+  // shared type parameters are equal. 
+  forall(int i | i in [0 .. -1 + min(int num | num = method.getNumTypeParameter() or num = other.getNumTypeParameter())] |
+    method.getTypeParameterBound(i) = other.getTypeParameterBound(i)
+  )
+}
+
+/**
+ * Holds if the two method-signatures are overloads of each other and have the same parameter types.
+ */
+predicate signaturesMatch(MethodSignature method, MethodSignature other) {
+  // declared in the same interface/class.
+  method.getDeclaringType() = other.getDeclaringType() and
+  // same static modifier.
+  getKind(method) = getKind(other) and
+
+  // same name.
+  exists(string name |
+    method = method.getDeclaringType().getMethod(name) and
+    other = other.getDeclaringType().getMethod(name)
+  ) and
+
+  // same number of parameters.
+  method.getBody().getNumParameter() = other.getBody().getNumParameter() and
+  
+  // The types are compared in matchingCallSignature. This is sanity-check that the textual representation of the type-annotations are somewhat similar.
+  forall(int i | i in [0 .. -1 + method.getBody().getNumParameter()] |
+    getParameterTypeAnnotation(method, i) = getParameterTypeAnnotation(other, i)   
+  ) and
+  
+  matchingCallSignature(method.getBody().getCallSignature(), other.getBody().getCallSignature())
+}
+
+from ClassOrInterface decl, string name, MethodSignature previous, MethodSignature unreachable
+where
+  previous = decl.getMethod(name) and
+  unreachable = getOtherMatchingSignatures(previous) and
+  
+  // If a later method overload has more type parameters, then that overload can be selected by explicitly declaring the type arguments at the callsite.
+  // This comparison removes those cases.
+  unreachable.getBody().getNumTypeParameter() <= previous.getBody().getNumTypeParameter() and
+  
+  // We always select the first of the overloaded methods.
+  not exists(MethodSignature later | later = getOtherMatchingSignatures(previous) |
+    previous.getFirstToken().getLocation().endsAfter(later.getFirstToken().getLocation())
+  )
+select unreachable,
+  "The latter overloads of " + name +
```

Comment:
The alert message reads a little awkwardly.

Reply:
I've changed it to:   
`"This overload of " + name + "() is unreachable, the $@ overload will always be selected.", previous, "previous"`

But I'm not that happy about either of them, so suggestions are welcome.